### PR TITLE
Lo Sharing v3

### PR DESCRIPTION
**Pull request by mdwhitten**
Simple version of LO sharing post internal discussions about the approach.

### CODE REVIEW

**Comment by mdwhitten** [comment]
@manderseck / @smooresni : Please check out commit 7e82834683054e98b6427bdd77e6eaa34f008122. I discovered a problem with implementing the `StoreAutomaticSGSAShared` function at the Instrument level. When it is called, it resets the upconverter mode. This is a problem for the 5840 if it is called _after_ calling `SetScriptToGenerate`, because it kicks the device out of IF mode. Hence, the LO offset is slightly wrong. I had to modify the TestStand sequence so that `ConfigureInstrument` is called after calling the configure generation mode settings.

This is probably how it should have been done all along, but I am specifically interested in feedback if the approach I took in the referenced commit (checking if set and not touching it if so) is a good idea. It doesn't solve the problem but it could help avoid making more problems.

**Comment by smooresni** [comment]
> @manderseck / @smooresni : Please check out commit [7e82834](https://github.com/NISystemsEngineering/rfic-reference-design-libraries-csharp/commit/7e82834683054e98b6427bdd77e6eaa34f008122). I discovered a problem with implementing the `StoreAutomaticSGSAShared` function at the Instrument level. When it is called, it resets the upconverter mode. This is a problem for the 5840 if it is called _after_ calling `SetScriptToGenerate`, because it kicks the device out of IF mode. Hence, the LO offset is slightly wrong. I had to modify the TestStand sequence so that `ConfigureInstrument` is called after calling the configure generation mode settings.
> 
> This is probably how it should have been done all along, but I am specifically interested in feedback if the approach I took in the referenced commit (checking if set and not touching it if so) is a good idea. It doesn't solve the problem but it could help avoid making more problems.

I wonder if R&D is aware of this atomic nature. I think what you are doing now is fine with the check and set, as that will solve most cases where this could arise. If the user does some weird toggling of the SG SA shared then calls initiate it will fail, but i don't foresee that happening often.

File: Source/SG/SG.cs [hunk 1]
```diff
@@ -104,18 +123,26 @@ public static void ConfigureInstrument(NIRfsg rfsgHandle, InstrumentConfiguratio
...
-                rfsgHandle.RF.LocalOscillator.Source = RfsgLocalOscillatorSource.Onboard;
+                IntPtr rfsgPtr = rfsgHandle.GetInstrumentHandle().DangerousGetHandle();
+                NIRfsgPlayback.RetrieveAutomaticSGSASharedLO(rfsgPtr, "", out RfsgPlaybackAutomaticSGSASharedLO currentMode);
+                if (instrConfig.LOSharingMode == LocalOscillatorSharingMode.None && currentMode != RfsgPlaybackAutomaticSGSASharedLO.Disabled)
+                {
+                    //Setting this property resets other settings, which can create issues. Hence, it is only set if the value
+                    //is different than the current mode.
+                    NIRfsgPlayback.StoreAutomaticSGSASharedLO(rfsgPtr, "", RfsgPlaybackAutomaticSGSASharedLO.Disabled);
+                }
+                else if (currentMode != RfsgPlaybackAutomaticSGSASharedLO.Enabled)
```
Comment (smooresni) [inline Source/SG/SG.cs]:
```suggestion
                else if (instrConfig.LOSharingMode == LocalOscillatorSharingMode.Automatic && currentMode != RfsgPlaybackAutomaticSGSASharedLO.Enabled)
```

Found during unit testing. If the LOSharingMode is set to None and currentMode is Disabled, current code will enable AutoSGSAShared. This suggestion should provide a stronger check for toggling the value.

**Comment by smooresni** [comment]
by the way, this will pass as soon as this change is made
https://github.com/NISystemsEngineering/rfic-reference-design-libraries-csharp/pull/94#discussion_r342789156

File: Source/SA/RFmx Instr/Properties/AssemblyInfo.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+// General Information about an assembly is controlled through the following
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("RFmx SpecAn")]
+[assembly: AssemblyDescription("")]
+[assembly: AssemblyConfiguration("")]
+[assembly: AssemblyProduct("RFmx SpecAn")]
```
[the remark this answers is not in the document]
Comment (mdwhitten) [reply Source/SA/RFmx Instr/Properties/AssemblyInfo.cs]: Yes; I believe this was your addition 😉 but I didn't catch it! Will fix.

File: Source/SA/RFmx Instr/RFmxInstr.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+            instrHandle.GetInstrumentModel("", out string model);
+            // Only configure LO settings on supported VSTs
+            if (Regex.IsMatch(model, "NI PXIe-58[34].")) // Matches 583x and 584x VST families
+            {
+                if (instrConfig.LOSharingMode == LocalOscillatorSharingMode.None)
+                    instrHandle.ConfigureAutomaticSGSASharedLO("", RFmxInstrMXAutomaticSGSASharedLO.Disabled);
+                else
+                    instrHandle.ConfigureAutomaticSGSASharedLO("", RFmxInstrMXAutomaticSGSASharedLO.Enabled);
+                // Configure automatic LO offsetting
+                instrHandle.SetLOLeakageAvoidanceEnabled("", RFmxInstrMXLOLeakageAvoidanceEnabled.True);
```
[the remark this answers is not in the document]
Comment (smooresni) [reply Source/SA/RFmx Instr/RFmxInstr.cs]: After speaking in person, we've decided to hardcode this value to true and reset the downconverter frequency and downconverter frequency offset properties

**Pull request merged**